### ISSUE TITLE
Question about raddstoreexpminusmax_ukernel

### ISSUE THREAD

**Issue by rohit-tangibleplay** [google/XNNPACK#1003]
Hi, 

I am running a tflite model which has a final softmax layer whose input is a heatmap of dimension 1x64x64x3 where 3 is the number of channels. The output dimension is 1x64x64. Tflite is built on mac and I use the c api with XNN delegate. When running this on an iPad, i get a EXC_BAD_ACCESS error and the program crashes.

I was able to narrow the error down to _'raddstoreexpminusmax_ukernel'_ function inside _'xnn_compute_f32_three_pass_softmax'_. Based on the device, the function _'xnn_f32_raddstoreexpminusmax_ukernel__neonfma_lut64_p2_x16'_ is called.

Within this function, the crash happens at the last call to 'xnn_compute_f32_three_pass_softmax' on line 263 of function '_xnn_f32_raddstoreexpminusmax_ukernel__neonfma_lut64_p2_x16_' which is
const float32x4_t vi = vld1q_f32(input); input += 4;

The two things i want to highlight are that the crash occurs at the last call of '_xnn_compute_f32_three_pass_softmax_' (ie. batch_index = 4095, total = 64x64 calls) and the value for variable '_elements_' inside _raddstoreexpminusmax_ukernel_ is 12 ie. 3 (number of channels) * sizeof(float)

From my observation, in the above line we are loading 4 inputs. However, in my case, the number of inputs is 3. My question is would this lead to out of bounds read specially for the last iteration of the function?

Thank you!

**Comment by Maratyszcza** [comment]
This does lead to out-of-tensor-bounds read for the last 4 elements, but TFLite arena allocator normally adds extra bytes in the end of allocated regions so the memory is within the allocated arena. I'll look into why this didn't work here.

**Comment by rohit-tangibleplay** [comment]
Hi @Maratyszcza , thanks for taking this up. I further probed the memory allocations inside _runtime.c_ and was able to find an interesting observation which I believe to be the reason for the crash. 

I checked the start and end addresses for each blob inside the for loop after '_xnn_allocate_simd_memory_' is called in _xnn_create_runtime_v2_ inside runtime.c. I get the following for the final softmax layer

```
printf("workspace data: %d, %p, %p, %d\n", i, blob->data, blob->data+blob->size, blob->size); 
>> workspace data: 107, 0x12b1c0000, 0x12b1cc000, 49152 //Softmax allocation
```

The size of the memory allocated is exactly equal to the size of the output tensor (ie. 64x64x3xsizeof(float)). The value of argument '_input_' inside '_xnn_f32_raddstoreexpminusmax_ukernel__neonfma_lut64_p2_x16_' during the last iteration of '_xnn_compute_f32_three_pass_softmax_' is 0x12b1cbff4 before crashing. I think since this being the last layer there is no additional padding elsewhere and when '_vld1q_f32_' tries to read 4 floats from 0x12b1cbff4 with the upper limit at 0x12b1cc000, the read goes out of bounds.

To understand why there was no extra padding added, I looked into the arguments of the function '_xnn_add_value_allocation_tracker_' couple of lines before simd_memory allocation. One of the arguments to that method is '_round_up_po2_' the arguments of which are blob size and XNN_EXTRA_BYTES. 

In our case, the blob size that is input to '_round_up_po2_' is 64 * 64 * 3 * sizeof(float) with XNN_EXTRA_BYTES = 16 (default). The output of this method is same as the input 64 * 64 * 3 * sizeof(float) and no extra bytes is added. I have added a few cases of '_round_up_po2_' outputs for different n and q below for reference

```
n: 64000, q: 16, round_up_po2: 64000
n: 12884, q: 16, round_up_po2: 12896
n: 49152, q: 16, round_up_po2: 49152
n: 49152, q: 64, round_up_po2: 49152
```

To summarize, I believe the out of bounds read is due to this being the last layer and round_up_po2 not adding any extra padding. I apologize for the long comment and please let me know if something is not clear.

Thanks!

**Comment by Maratyszcza** [comment]
Thank you for the details. This bug was presumably fixed in XNNPACK in 68447302abcfad0d4b6b19a1efe7d7eef8833f4a and in TensorFlow Lite in tensorflow/tensorflow@cc0621d70fb6c3d4e4b3b92d65f1c8fdfafe014f. Please let me know if you continue to see the crash.
